Emit dns error thrown by native code

achingbrain opened the pull request:
The native `dns_sd.DNSServicePRocessResult` method can [throw an error](https://github.com/agnat/node_mdns/blob/7832eecc07d49f49735c771b1e00e06e513d1067/src/dns_service_process_result.cpp#L26) if there's a DNS failure, resulting in an uncaught exception and your application going down, so this pull request wraps it in a try/catch block and gets the `Browser` object (extends `MDNSService`) to emit an error event instead.

Also includes a unit test.  Actually the `test_functional`/`simple browsing` test is failing at the moment but I figure better fixed in a different commit.

agnat commented:
Thanks! Reviewing...

agnat commented on lib/browser.js:
```diff
@@ -36,9 +36,19 @@ var Browser = exports.Browser = function Browser(serviceType, options) {
+        self.emit('error', e)
+        return
+      }
+
```
Please use semicolons after statements.

agnat commented on package.json:
```diff
@@ -19,7 +18,7 @@
 , "repository":
   { "type": "git"
-  , "url":  "http://github.com/agnat/node_mdns.git"
+  , "url":  "http://github.com/achingbrain/node_mdns.git"
```
Please undo the changes to package.json. They do not belong into the PR. Note, that you don't need to republish a package just to get your PR in. Use the tarball or git URL of your PR in _your_ package.json. Much less noisy... See the last paragraph of the [Installation Section](https://github.com/agnat/node_mdns#installation) of the README for an example.

agnat commented on tests/test_browser.js:
```diff
@@ -68,5 +68,31 @@ module.exports["Browser"] = {
+    };
+
+    // create the browser with our mock of dns_sd
+    var browser = proxyquire('../lib/browser.js', { './dns_sd': mock_dns_sd }).Browser.create(serviceName);
```
`proxyquire(...)` pops up out of nowhere? No, `require('proxyquire')`? How does that work?

achingbrain replied:
It's been there for a while:

https://github.com/agnat/node_mdns/blob/2ff8d02505198e4e33de30b823e9ec65b06c2837/tests/test_browser.js#L3

agnat commented on tests/test_mdns_service.js:
```diff
@@ -0,0 +1,29 @@
+var dns_sd = require('../lib/dns_sd')
+  , proxyquire =  require('proxyquire')
```
More proxyquire business. Here it is required. But it doesn't appear in the `devDependencies` in package.json. Is it installed globally on your system?

achingbrain replied:
No, not globally installed.  It's in devDependencies:

https://github.com/agnat/node_mdns/blob/e38418c22ffc8d7db09e1bc689f617a4f3c6e9a5/package.json#L18

agnat replied:
Hehe... apparently I have no idea what's going on in my software. So, thanks for the explanation.

agnat commented:
I finished my review.

Still not sure about proxyquire, though. What it does is quite neat, no question. But it sounds like it is quite a messy business. I will probably give that a glance too. Do you have long-term experience in using it? How stable is it (across different node versions)?

achingbrain commented:
Sorry for my delay in replying - I've somehow managed to rent a holiday apartment without WiFi.  Schoolboy error.

Sorry also for the semi-colon thing - I'd mean to push the subsequent commit to a branch instead of master so didn't mean those changes to be part of this pull request.  They do appear to be valid though - when using node-mdns on really flaky WiFi networks I noticed service names would occasionally come back as an empty string.  I've added the extra punctuation.

As for the publishing to npm/using a git url/tarball - I've been deploying in corporate environments with silly content filter policies so sometimes github is blocked but npm isn't.  Obviously in this case using github URLs won't work so it's easier to just publish the fork under a different name then delete it once pull requests are merged to the upstream version and new versions are released.  Anyway I've rebased that commit out of the repo history.

Proxyquire - I've used it quite a bit.  You are right it does some horrible things to `require` but it gets the job done.  It gets 80k+ downloads a month on npm so you'd hope it's stable ;)

That said I'd rather see some sort of dependency injection so it's not necessary but that would turn into quite the rewrite.

agnat commented:
> Sorry for my delay in replying [...] Sorry also for the semi-colon thing. [...] 

No worries.

The patch looks good to me... after resolving my confusion about proxyquire. ;)

@ronkorving, care to take a second look and release the fix?

ronkorving commented:
Looks good to me, thanks :)

ronkorving commented:
Released
